packages/perseus-core/src/parse-perseus-json/perseus-parsers/radio-widget.test.ts

```diff
@@ -108,6 +108,193 @@ describe("parseRadioWidget", () => {
+                    ],
+                },
+            };
+
+            const result = parse(widget, parseRadioWidget);
+
+            expect(result).toEqual(
+                success({
+                    ...widget,
+                    options: {
+                        ...widget.options,
+                        choices: [
+                            {
+                                content: "Choice without ID",
+                                id: "radio-choice-0", // Should be generated
+                            },
+                        ],
+                    },
+                }),
+            );
+        });
+
+        it("generates default ID when choice ID is null", () => {
+            const widget = {
+                type: "radio",
+                version: {major: 3, minor: 0},
+                graded: true,
+                options: {
+                    choices: [
+                        {
+                            content: "Choice without ID",
+                            id: null,
+                        },
+                    ],
+                },
+            };
+
+            const result = parse(widget, parseRadioWidget);
+
+            expect(result).toEqual(
+                success({
+                    ...widget,
+                    options: {
+                        ...widget.options,
+                        choices: [
+                            {
+                                content: "Choice without ID",
+                                id: "radio-choice-0", // Should be generated
+                            },
+                        ],
+                    },
+                }),
+            );
+        });
+
+        it("generates correct IDs for multiple choices", () => {
+            const widget = {
+                type: "radio",
+                version: {major: 3, minor: 0},
+                graded: true,
+                options: {
+                    choices: [
+                        {content: "First choice", id: ""},
+                        {content: "Second choice", id: "custom-id"},
+                        {content: "Third choice", id: "  "},
+                        {content: "Fourth choice"}, // undefined ID
+                    ],
+                },
+            };
+
+            const result = parse(widget, parseRadioWidget);
+
+            expect(result).toEqual(
+                success({
+                    ...widget,
+                    options: {
+                        ...widget.options,
+                        choices: [
+                            {content: "First choice", id: "radio-choice-0"},
+                            {content: "Second choice", id: "custom-id"}, // preserved
+                            {content: "Third choice", id: "radio-choice-2"},
+                            {content: "Fourth choice", id: "radio-choice-3"},
+                        ],
+                    },
+                }),
+            );
+        });
+    });
```

Comment:
These tests cover all the cases, but they're a little verbose, and they're really testing `defaultedNonEmptyString`. I'd prefer to have comprehensive tests in `defaulted.test.ts` that call `defaultedNonEmptyString` directly, and then we'd only need to test a couple cases here (maybe the empty ID case and the nonempty ID case).

If there's published content that has blank choice IDs, we should also put an example of that in a regression test data file in `regression-tests/item-data`.